ingestion/src/levels.rs

```diff
@@ -0,0 +1,168 @@
+//! - `0 = default` is not what most end users want (although they got used to it)
+//!   meaning we either increase the burden on them, or convert to a different
+//!   scheme at request time, which is redundant and confusing.
+//! - `0 = default` is a friction point with our international collaborators
+//!   who nowadays aim to publish physical height instead.
+//!   Only Norway met and UK met have not moved away from this practice.
+//! - `NULL = 0` removes the ability to use NULL for cases where we don't know
+//!   the level, or it isn't relevant.
+//! - Inconsistent units and signed integers for level are likely a result of
+//!   the system originally being designed for positive integer metres only,
+//!   and then having to tack on sub-metre and negative levels.
+//!   Might as well fix that while we're making changes.
+//!
+//! These semantics do not apply to levels in the `kvalobs` and `obsinn` labels
+//! as those source-specific labels are meant to reflect how the data was
+//! reported in to lard, which in those cases will keep the old semantics.
+use std::{
+    collections::HashMap,
+    sync::{Arc, RwLock},
+};
+use thiserror::Error;
+use tokio_postgres::NoTls;
+use tracing::{error, warn};
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("postgres returned an error: {0}")]
+    Database(#[from] tokio_postgres::Error),
+    #[error("RwLock was poisoned: {0}")]
+    Lock(String),
+    #[error("issues with level conversion: {0}")]
+    Level(String),
+}
+
+#[derive(Debug, Clone)]
+pub struct Level {
+    hlevel: i32,
+    hlevel_scale: Option<i32>,
+}
+
+#[cfg(feature = "integration_tests")]
+impl Level {
+    pub fn new(hlevel: i32, hlevel_scale: i32) -> Level {
+        Level {
+            hlevel,
+            hlevel_scale: Some(hlevel_scale),
+        }
+    }
+}
+
+type ParamID = i32;
+
+/// this table is where to look for the default level and scale
+/// for a given parameter
+pub type ParamLevelTable = HashMap<ParamID, Level>;
+
+/// Get a fresh cache of levels from stinfosys
+pub async fn fetch_levels(stinfo_conn_string: &str) -> Result<ParamLevelTable, Error> {
+    // get stinfo conn
+    let (client, conn) = tokio_postgres::connect(stinfo_conn_string, NoTls).await?;
+
+    // conn object independently performs communication with database, so needs it's own task.
+    // it will return when the client is dropped
+    tokio::spawn(async move {
+        if let Err(e) = conn.await {
+            error!("connection error: {}", e);
+        }
+    });
+
+    // query param table
+    let rows = client
+        .query(
+            "SELECT standard_hlevel, hlevel_scale, paramid 
+                FROM param WHERE standard_hlevel is not null",
+            &[],
+        )
+        .await?;
+
+    // build hashmap of param permits
+    let mut param_level = HashMap::new();
+
+    for row in rows {
+        let hlevel_scale: i32 = row.get(1);
+        if ![0, -2].contains(&hlevel_scale) {
```

Comment:
@lcoram @Lun4m There is a bug here. The `!` should not be there. Might be a good idea to test this function before merging the next PR

Reply:
fix in #113